promotion.md

```diff
@@ -111,6 +118,41 @@ it is unlikely to be the desired behavior in most cases and very likey to be con
 to the user. If such behavior is desired, the user can still use an explicit `static`
 or `const` item and refer to that.
 
+*Dynamic check.* The Miri engine could dynamically check this by ensuring that
+ the result of computing a promoted is a value that does not need dropping.
+
+## `&` in `const` and `static`
+
+Promotion is also responsible for making code like this work:
+
+```rust
+const FOO: &'static i32 = {
+    let x = &13;
+    x
+};
+```
+
+However, since this is in explicit const context, we could be less strict about
+promotion in this situation.
+
+Promotion is *not* involved in something like this:
+
+```rust
+#![feature(const_vec_new)]
+const EMPTY_BYTES: &Vec<u8> = &Vec::new();
+
+const NESTED: &'static Vec<u8> = {
+    // This does not work when we have an inner scope:
+    let x = &Vec::new(); //~ ERROR: temporary value dropped while borrowed
+    x
+};
+```
+
+In `EMPTY_BYTES`, the reference obtains the lifetime of the "enclosing scope",
+similar to how `let x = &mut x;` creates a reference whose lifetime lasts for
+the enclosing scope. This is decided during MIR building already, and does not
+involve promotion.
```

Comment:
> I don't know how it should extend to arbitrary cases. Is it the braces that cause the problem or the assignment to x?

I don't know. I am paraphrasing @eddyb here.
@eddyb are these rules explained anywhere?

> It also might be good to mention #[rustc_promotable] here, since without it calls to const fns can't get promoted.

Good point, will do.

Reply:
Hm I guess I should edit this in the light of our unresolved terminology.

Is the entire "Promotion is *not* involved" incorrect, or is there something there that makes sense? I was trying to explain why `const EMPTY_BYTES: &Vec<u8> = &Vec::new();` works even though this stuff does *not* get promoted. @ecstatic-morse since my explanation confused you, could you try to explain this in your own words?

Reply:
See above comment for what is actually confusing me here. The "enclosing scope" rule makes sense intuitively, but the details are still not clear to me. That's my problem though, not this PR's.

Reply:
I have done some edits. @ecstatic-morse does this resolve your confusion?

Reply:
The enclosing scope rule is the same one that makes these two different:
```rust
{
    let r = &foo(); // compiles
    bar(r);
}
```
```rust
{
    let r = id(&foo()); // doesn't compile
    bar(r);
}
```
Except instead of `let` it's `const`/`static` and instead of "enclosing block" it's `'static`.

I don't remember the exact name of these semantics, maybe "rvalue/temporary lifetime/scope"? @nikomatsakis would be more helpful here, sorry.

Reply:
I call these "temporary lifetimes" -- I've been meaning to writeup some text about them for the rust reference.

Reply:
I call these "temporary lifetimes" -- I've been meaning to writeup some text about them for the rust reference.